src/repr/src/adt/datetime.rs

```diff
@@ -1905,11 +1905,75 @@ pub(crate) fn split_timestamp_string(value: &str) -> (&str, &str) {
     }
 }
 
+/// An encoded packed variant of [`NaiveTime`].
+///
+/// We uphold the invariant that [`PackedNaiveTime`] sorts the same as
+/// [`NaiveTime`].
+#[derive(Copy, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct PackedNaiveTime([u8; Self::SIZE]);
+
+impl PackedNaiveTime {
+    pub const SIZE: usize = 8;
+
+    /// Returns a reference to the underlying bytes.
+    pub fn as_bytes(&self) -> &[u8; Self::SIZE] {
+        &self.0
+    }
+
+    /// Interprets a slice of bytes as a [`PackedNaiveTime`].
+    ///
+    /// Returns an error if the size of the slice is incorrect.
+    pub fn from_bytes(slice: &[u8]) -> Result<Self, String> {
```

Comment:
This could be implemented a bit more directly via `try_into`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b3c429d443727dbe68bc28817f9aea9d

I wasn't sure how I felt about this leaving data only partially validated, but I think it's fine to panic at read time. Hard to imagine we'd ever do anything else in production...

Reply:
Ahhh nice! updated to use `try_into`